src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -170,21 +170,23 @@ public bool HasResponseStarted
             get { return _requestProcessingStatus == RequestProcessingStatus.ResponseStarted; }
         }
 
-        protected FrameRequestHeaders FrameRequestHeaders => _frameHeaders.RequestHeaders;
+        protected FrameRequestHeaders FrameRequestHeaders { get; private set; }
+
+        protected FrameResponseHeaders FrameResponseHeaders { get; private set; }
 
         public void InitializeHeaders()
         {
-            if (_frameHeaders == null)
+            if (FrameRequestHeaders == null)
             {
-                _frameHeaders = new Headers(ServerOptions);
-                RequestHeaders = _frameHeaders.RequestHeaders;
-                ResponseHeaders = _frameHeaders.ResponseHeaders;
+                RequestHeaders = FrameRequestHeaders = new FrameRequestHeaders();
```

Comment:
I'm not sure it gets much better. You'd get into issues with the explicit implementations of `IDictionary<string, StringValues>` in `FrameHeaders` and casting in Frame.FeatureCollection. Example:

```csharp
IHeaderDictionary IHttpRequestFeature.Headers
{
    get
    {
        return RequestHeaders;
    }

    set
    {
        // This won't work if RequestHeaders is FrameRequestHeaders. Cast?
        RequestHeaders = value;
    }
}
```